File: src/main/java/waterpower/common/network/PacketUnitChanged.java
package waterpower.common.network;

import io.netty.buffer.ByteBuf;
import net.minecraft.tileentity.TileEntity;
import net.minecraft.util.math.BlockPos;
import net.minecraft.world.World;
import net.minecraftforge.common.DimensionManager;
import net.minecraftforge.fml.common.network.simpleimpl.IMessage;
import net.minecraftforge.fml.common.network.simpleimpl.IMessageHandler;
import net.minecraftforge.fml.common.network.simpleimpl.MessageContext;
import waterpower.common.block.tileentity.IUnitChangeable;

public class PacketUnitChanged implements IMessage, IMessageHandler<PacketUnitChanged, IMessage> {
    private int dim, unitId;
    private long pos;

    public PacketUnitChanged() {
    }

    public PacketUnitChanged(int dim, BlockPos pos, int unitId) {
        super();
        this.dim = dim;
        this.pos = pos.toLong();
        this.unitId = unitId;
    }

    @Override
    public IMessage onMessage(PacketUnitChanged message, MessageContext ctx) {
        World world = DimensionManager.getWorld(message.dim);
        TileEntity te = world.getTileEntity(BlockPos.fromLong(message.pos));
        if (te instanceof IUnitChangeable) {
            IUnitChangeable tebg = (IUnitChangeable) te;
            tebg.setUnitId(message.unitId);
        }
        return null;
    }

    @Override
    public void fromBytes(ByteBuf buf) {
        dim = buf.readInt();
        pos = buf.readLong();
        unitId = buf.readInt();
    }

    @Override
    public void toBytes(ByteBuf buf) {
        buf.writeInt(dim);
        buf.writeLong(pos);
        buf.writeInt(unitId);
    }

}

File: src/main/kotlin/waterpower/common/network/PacketUnitChanged.kt
/**
 * Copyright (c) Huang Yuhui, 2017
 *
 * "WaterPower" is distributed under the terms of the Minecraft Mod Public
 * License 1.0, or MMPL. Please check the contents of the license located in
 * http://www.mod-buildcraft.com/MMPL-1.0.txt
 */
package waterpower.common.network

import io.netty.buffer.ByteBuf
import net.minecraft.util.math.BlockPos
import net.minecraftforge.common.DimensionManager
import net.minecraftforge.fml.common.network.simpleimpl.IMessage
import net.minecraftforge.fml.common.network.simpleimpl.MessageContext
import net.minecraftforge.fml.relauncher.Side
import waterpower.annotations.Message
import waterpower.api.IUnitChangeable
import waterpower.common.Energy

@Message(Side.SERVER)
class PacketUnitChanged() : Packet<PacketUnitChanged>() {
    private var dim: Int = 0
    private var energy: Energy = Energy.EU
    private var pos: Long = 0

    constructor(dim: Int, pos: BlockPos, energy: Energy) : this() {
        this.dim = dim
        this.pos = pos.toLong()
        this.energy = energy
    }

    override fun onMessage(message: PacketUnitChanged, ctx: MessageContext): IMessage? {
        val world = DimensionManager.getWorld(message.dim)
        val te = world.getTileEntity(BlockPos.fromLong(message.pos))
        if (te is IUnitChangeable)
            te.setUnit(message.energy)
        return null
    }

    override fun fromBytes(buf: ByteBuf) {
        dim = buf.readInt()
        pos = buf.readLong()
        energy = Energy.values()[buf.readInt()]
    }

    override fun toBytes(buf: ByteBuf) {
        buf.writeInt(dim)
        buf.writeLong(pos)
        buf.writeInt(energy.ordinal)
    }

}
